### ISSUE TITLE
mirage not working for mocking ember ajax calls

### ISSUE THREAD

**Issue by rakesh9** [samselikoff/ember-cli-mirage#635]
I am new to mirage and in ember application, I would like to use mirage for mocking ember ajax calls but it is not working.
is mirage works only with ember data? I didn't find any example to mock ember ajax calls.

Below are my snippets..

making a ember ajax call from component and defined corresponding route in mirage/config.js and stated the mirage in my integration tests.
when I make ajax the call is being going to mocked route in config.js but it is not returning any promise back to the caller and none of callback functions done, always, failed were called.
## Do i missing something???

/mirage/config.js
export default function() {
 this.get('customers/details', function(db, request) {
    console.log("inside customer ajax call");

```
return new Mirage.Response(400, { some: 'header' }, { message: 'name cannot be blank' });
```

  });
}
/component.js

Ember.$.ajax({
      url: 'customers/details',
      type: 'get',
      dataType: 'json',
      async: false,
      beforeSend: function (request) {
        console.log("before req");
      },
      readyStateChanged: function (xhr) {
       if (xhr.readyState == 1) {
         console.log("state chnaged");
       }
     }
    }).done(function (result) {
      console.log("done"+result);
    }).fail(function (jqXHR) {
      console.log("failed");
    }).always(function (jqXHR) {
      console.log("always");
    });

---

/component-test.js

import Ember from 'ember';
import { moduleForComponent, test } from 'ember-qunit';
 import startMirage from '../../../helpers/setup-mirage-for-integration';
 import hbs from 'htmlbars-inline-precompile';

 moduleForComponent('customer-details', 'Integration | Component | customer details', {
   integration: true,
   setup: function() {
     startMirage(this.container);
   }
 });
 test('it renders', function(assert) {
   expect(0);
   console.log("test");
 this.render(hbs`{{customer-details}}`);
 });

---

**Comment by samselikoff** [comment]
Nope, Mirage works for any XHR request, so Ember Data is not the issue.

Do you never see that `console.log("inside customer ajax call");`? Try putting a `debugger` within your Mirage route handler:

```js
this.get('customers/details', function(db, request) {
  debugger;

  return new Mirage.Response(400, { some: 'header' }, { message: 'name cannot be blank' });
});
```

Does that ever get called? Does it work in development?

**Comment by rakesh9** [comment]
I see console.log("inside customer ajax call") is being executing..it means the mirage is able to route my request to this mock but seems like it is not returning any promise and none of my callback methods done, fail, always being called.

**Comment by samselikoff** [comment]
Yep, that means Mirage has intercepted it. What about putting a `debugger` in `always`?

```js
Ember.$.ajax({
  url: 'customers/details',
  type: 'get',
  dataType: 'json'
}).always(function() {
  debugger;
});
```

**Comment by rakesh9** [comment]
I tried but always function is not being called. This is the issue that I have been facing..mirage is intercepting my ajax req but not returning json response and none of callback methods including always never called

**Comment by samselikoff** [comment]
Does the console have any errors or anything?

Could you upload a sample repo on github demonstrating the problem?

**Comment by rakesh9** [comment]
doesn't have any errors in console...I am adding mirage to the existing project. let me see If I can extract relevant files and  put it in githup

**Comment by jeremy-w** [comment]
I've run into this as well. Took a lot of poking around to localize it to adding Mirage.

When I stepped through, it looked like the state was never getting properly updated on those fake/wrapped XHRs, but I'm not certain of that due to some convoluted code. Ultimately, the jQuery Deferred backing $.ajax never gets fulfilled, so everything hangs forever waiting on a callback that never comes.

Even weirder, it affected only the second test in the acceptance test module; all the XHRs in the second test would just never resolve or reject.

I haven't boiled it down to any reasonably-sized reproducible case yet, though.

**Comment by jeremy-w** [comment]
@samselikoff I have reduced the app in which I encountered this issue to a minimal skeleton that still exhibits the issue. Please clone [this repository](https://github.com/jeremy-w/mirage-xhr-hanging-repro) and build and run in order to reproduce:

```
git clone https://github.com/jeremy-w/mirage-xhr-hanging-repro.git
cd mirage-xhr-hanging-repro
npm install
bower install
ember s
```

Then visit [the tests page](http://localhost:4200/tests) to see the tests hang.

You can flip a config variable to flip Mirage off, and watch them go through just fine, the flip Mirage back on for tests, and watch it hang again. It's weird stuff.

**Comment by jeremy-w** [comment]
Update: This issue has been resolved in my case.

- The issue I was seeing with Mirage in the app was due to an older Pretender version still being pulled in by Bower. Running `ember g ember-cli-mirage` and then `bower install` and picking the new versions for any resolutions fixed the issue.
- The issue I was seeing in my minimization case was due to the return values of my factories. In both the Mirage and raw PretenderJS branches, the return value was tripping an exception, which interrupted the XHR state machine. This exception was silently swallowed somewhere in the stack (jQuery's stuff?), which made it appear the same as the other issue symptomatically. Trapping on `debugger` in the handler, then flipping on "stop on exceptions" and "yes really all of them even caught ones" allowed to discover the error that should have been shown and adjust the responses, at which point the tests passed.

**Comment by samselikoff** [comment]
@jeremy-w glad you finally figured it out, sorry it took so long & I wasn't able to jump in :/ 

The exception swallowing can be brutal. I tried to err on the side of over-throwing errors in Mirage for these situations (which is why sometimes you'll see the same error show up in the console multiple times), but sometimes this sort of thing still seems to happen.
